daft/dataframe/dataframe.py

```diff
@@ -1044,6 +1011,53 @@ def collect(self, num_preview_rows: Optional[int] = 8) -> "DataFrame":
 
         return self
 
+    @DataframePublicAPI
+    def show(self, n: int = 8) -> "DataFrameDisplay":
+        """Executes enough of the DataFrame in order to display the first ``n`` rows
+
+        .. NOTE::
+            This call is **blocking** and will execute the DataFrame when called
+
+        Args:
+            n: number of rows to show. Defaults to 8.
+
+        Returns:
+            DataFrameDisplay: object that has a rich tabular display
+        """
+        preview_partition = self._preview.preview_partition
+        total_rows = self._preview.dataframe_num_rows
+        if preview_partition is None or (
+            n > len(preview_partition) and (total_rows is None or len(preview_partition) < total_rows)
```

Comment:
Why is `(total_rows is None or len(preview_partition) < total_rows)` necessary?

IIRC total_rows is just used to display the message `Showing <N> of <total_rows> rows`, but `.show()` doesn't show the total rows anyways so shouldn't this be sufficient?

```python
if preview_partition is None or or n > len(preview_partition):
    # materialize preview partition
elif n < len(preview_partition):
    # slice preview partition
else:
    assert n == len(preview_partition)
    # return existing preview partition
```

Reply:
This is an obvious optimization that handles the case when the length of the full DataFrame is less than `n`, in which case the DataFrame preview will have length than `n` but we'll still want to use the cached preview, since it already consists of the entire `DataFrame`.

E.g. if we have a DataFrame with 5 total rows that we've already materialized with `df.collect()`, and we then call `df.show()`, we'll have a preview of the full DataFrame with 5 rows; this is technically less than the default `show()` limit (8), but we obviously don't want to recompute the DataFrame in this case, since we already have the full DataFrame as a preview!

Reply:
I see! Small request on my end to invert the if/elses, which might help make the conditions easier to reason about

```python
has_preview_partition = preview_partition is not None

if has_preview_partition and len(preview_partition) == n:
    # Preview partition exists and is exactly the size we need so we return it
elif has_preview_partition and len(preview_partition) > n :
    # Preview partition exists, but is too large so we slice it
elif (
    has_preview_partition and
    total_rows is not None and
    n > total_rows and
    len(preview_partition) == total_rows
):
    # Preview partition exists and the requested n is larger than total_rows
    # If the preview partition is already all the rows , we can just return it.
else:
    # No choice but to materialize preview partition:
    # 1. has_preview_partition=False
    # 2. preview_partition exists, but is too small
```

Reply:
I ended up truncating `n` to `total_rows`, which made all of the logic a good bit more straightforward while keeping us at a single level of conditionals!